digest-macro-derive/src/lib.rs

```diff
@@ -0,0 +1,50 @@
+extern crate proc_macro;
+
+use crate::proc_macro::TokenStream;
+use quote::quote;
+use syn::{Data, Fields};
+
+#[proc_macro_derive(DigestRootMacro)]
+pub fn digest_root_macro_fn(input: TokenStream) -> TokenStream {
+    let ast = syn::parse(input).unwrap();
+    impl_digest_root_macro(&ast)
+}
+
+fn impl_digest_root_macro(ast: &syn::DeriveInput) -> TokenStream {
+    let name = &ast.ident;
+    if let Data::Struct(data) = &ast.data {
+        if let Fields::Named(fields) = &data.fields {
+            let idents = fields
+                .named
+                .iter()
+                .map(|field| field.ident.as_ref().expect("Named field"));
+
+            let types = fields.named.iter().map(|field| &field.ty);
+
+            let gen = quote! {
+            impl DigestRoot for #name
+                 where #(#types: DigestField),*
+                 {
+                    fn digest_root(self) -> Multihash {
+                        let mut digests = vec![];
+                        #(digests.push(self.#idents.digest_field(stringify!(#idents).into())););*
+
+                        digests.sort();
+
+                        let res = digests.into_iter().fold(vec![], |mut res, digest| {
+                            res.append(&mut digest.into_bytes());
+                            res
+                        });
+
+                        digest(&res)
+                    }
+                }
+            };
+            gen.into()
+        } else {
+            panic!("DigestRootMacro does not support new types.");
```

Comment:
I don't quite understand this error path.  This matches on `Fields` and is ok for `Fields::Named` so this error path is for `Fields::Unnamed` or `Fields::Unit`.  Perhaps this string could be:

"DigestRootMacro only supports named fields"

Reply:
"named field" did not mean much to me, so I wanted to use terminology I am more familiar such as `new types`. But indeed, the current message does not cover `tuples` and `unit types`.